Add GitHub Action deployment

fregante opened the pull request:
Closes #19 

Needs a bit more work and testing, but at least the overall logic seems to work

fregante committed:
Add `deploy` workflow

fregante committed:
Create deploy-daily.yml

fregante committed:
Disable tag publishing (temporary)

fregante committed:
Update package.json

fregante committed:
Fix deploy-daily.yml so it doesn't error

fregante committed:
Run generic tests on GHA

fregante committed:
Delete .travis.yml

fregante committed:
Fix deploy-daily.yml so it doesn't error /2

fregante committed:
Fix dependency + change prerelease scripts

fregante committed:
Group versioning scripts

fregante committed:
Add deployment secrets ENV

fregante commented:
I think this is ready, it just needs some documentation.

Also it would be great if you could send this extension to CWS and AMO for real, but unlisted, so we can make sure this always actually works _and_ we can keep the `yml` files _active_ instead of having to disable the push commands (like I did with `echo` and commenting them out)

fregante commented:
If you can publish this extension and plug in the secrets in your settings it'd be great so we can merge it.

sindresorhus commented on .github/workflows/deploy-daily.yml:
```diff
@@ -0,0 +1,21 @@
+  create-tag:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v1
```
```suggestion
    - uses: actions/checkout@v2
```

sindresorhus commented:
This has a lot of boilerplate. Instead of a template, why not make a reusable GitHub Actions action?

fregante committed:
Use actions/checkout@v2, drop setup-git-token

fregante committed:
Use actions/checkout@v2, drop setup-git-token

fregante commented:
It wouldn't save a whole lot if we want to keep the publishing to the 2 stores in 2 separate jobs (which would save us from a few issues we had before)

If you look at the actual code run, it's very specific to this workflow. Most of the boilerplate is because we can't easily share data between jobs (e.g. the version)

fregante commented:
It could be further shortened by merging `delete-tag`, `build` and `tag` jobs, but not substantially

fregante committed:
Temporarily use fixed `copy-webpack-plugin` version

They changed the schema and it's now failing

fregante committed:
Merge 3 jobs; fix tagging

fregante committed:
Merge version and tag

notlmn commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,13 @@
+  test:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v1
```
Why not `v2` here?

notlmn commented on package.json:
```diff
@@ -15,7 +15,7 @@
 	"devDependencies": {
 		"chrome-webstore-upload-cli": "^1.2.0",
-		"copy-webpack-plugin": "^5.0.3",
+		"copy-webpack-plugin": "5.0.3",
```
Let's fix this in this PR as per https://github.com/notlmn/browser-extension-template/issues/22#issuecomment-570781123, or will have to insert package-lock.json again? WDYT?

fregante replied:
I don't know how much more work is needed here, so it's best to fix that separately.

notlmn reviewed (changes requested):
Added the required GH secrets, created a new extension on CWS and AMO.

AMO: https://addons.mozilla.org/en-US/firefox/addon/awesome-extension/
CWS: https://chrome.google.com/webstore/detail/awesome-extension/bdeobgpddfaegbjfinhldnkfeieakdaf

(Links may not active immediately, created just now)

fregante committed:
actions/checkout@v2 in test.yml

notlmn commented:
![image](https://user-images.githubusercontent.com/37769974/74526654-d56b3000-4f49-11ea-83d4-29a2278792ad.png)

AMO reviewers don't let me list the extension, no word from CWS reviewers.

The publish part _should_ work as intended though, or I might be wrong.

fregante commented:
Sounds bad at first but it just says it can't be listed by them: that means we can still use the whole process but it won't appear on the site.

That's exactly what we need. We don't actually need to distributed it, but just to test the upload process.

fregante committed:
Change: tagger + tag deployer workflows

Instead of: cronjob + tagger/deployer workflow.

This enables users to also manually create specific version tag instad of just `deploy`

fregante committed:
Fix manifest versioning

fregante committed:
Merge branch 'master' into github-action

fregante commented on package.json:
```diff
@@ -8,15 +8,15 @@
 		"release:cws": "webstore upload --source=distribution --auto-publish",
-		"release:amo": "web-ext-submit --source-dir distribution"
+		"release:amo": "web-ext-submit --source-dir distribution",
+		"release": "VER=$(daily-version) run-s build version release:*"
```
The last improvement I'd make would be to let the user optionally set the version:

```sh
npm run release # uses daily-version
npm run release -- 20.1.1 # uses 20.1.1 as version
```

Suggestions welcome for this, but it can also come in the future.

notlmn replied:
How would the user specify the version, modify package.json in a commit?

notlmn replied:
This is similar to manually creating a hotfix tag right, or am I missing something?

fregante replied:
> How would the user specify the version, modify package.json in a commit?

Currently we never save the version in the repo itself, we just create tags. The use would specify the version when releasing, as specified: 
```sh
npm run release -- 20.1.1 # uses 20.1.1 as version
```

But this ⬆️  feature hasn't been added yet.

> This is similar to manually creating a hotfix tag right, or am I missing something?

Kinda, yeah, but `hotfix` is just a way to trigger GHA; if you want to deploy from your computer, you can just run the `release` npm script. 

Before this, there was no easy way to deploy from your computer because the scripts were tightly integrated with Travis.

notlmn replied:
Something like `VER=$([ -z $1 ] && daily-version || echo $1)` should let us do `npm run release 20.1.1`, right?

fregante replied:
That's not how npm script works, you can only pass arguments after `--`

`npm run release -- 20.3.25` probably won't work either. This can be a separate PR

fregante committed:
Revert unrelated dependency changes (169d1ff)

fregante commented on .github/workflows/deploy-on-version-tag.yml:
```diff
@@ -0,0 +1,57 @@
+    - run: cd distribution && echo npx web-ext-submit # TODO: drop `echo` before merging
+      env:
+        WEB_EXT_API_KEY: ${{ secrets.WEB_EXT_API_KEY }}
+        WEB_EXT_API_SECRET: ${{ secrets.WEB_EXT_API_SECRET }}
```
If you set these ENVs on Refined GitHub, we can drop Travis deployment (and its issues/quirks) and move entirely to GHA

notlmn commented:
> That's exactly what we need. We don't actually need to distributed it, but just to test the upload process.

Manually ran `web-ext-submit` to see if Mozilla is still allowing publishing extension, they do, but it unlisted and only owners can see the extension on AMO. Which is also kinda what we want but would be a problem if they remove it entirely at some point. Have to ask them to keep the extension unlisted and restricted if they do take it down in the future.

We do not have any issues with Chrome extension (apart from the very slow review cycles), the extension is unlisted and allows deployment without any complaints.

-------------

We should also update docs accordingly, moving all steps from Travis to GHA, and also mention how the new auto-publishing mechanism works.

notlmn commented on .github/workflows/deploy-on-version-tag.yml:
```diff
@@ -0,0 +1,57 @@
+    - uses: actions/checkout@v2
+    - run: npm install
+    - run: npm test
+    - run: du -a # Log all available files # TODO: maybe remove
```
I like `ls -ARlF` more, combined with `-I node_modules -I .git`, and any other directories that we do not want to list.

(Or `du -a --exclude='node_modules' --exclude='.git'`, whatever you prefer)

fregante replied:
This was only for debugging of this	action, we can drop it

fregante replied:
```suggestion
```

notlmn commented on .github/workflows/deploy-automatic.yml:
```diff
@@ -0,0 +1,28 @@
+# If new commits are available since the last tag, create a new version/tag
+# - Every day at 13:13
```
Why the magic time `13:13` though, just curious.

fregante replied:
Because it's a lucky number!

hehe

Any time works, as long as it's not the "default" midnight (`0 0 * * *`)

notlmn committed:
Upadate Gecko extension ID in `manifest.json`

fregante commented:
> We should also update docs accordingly, moving all steps from Travis to GHA, and also mention how the new auto-publishing mechanism works.

Will do

notlmn commented:
![image](https://user-images.githubusercontent.com/37769974/77547563-5ca9ad00-6ed3-11ea-90ea-94ea2e0f3501.png)

Got another notice from AMO reviewers, maybe we should use a stub for AMO deployment.

fregante commented:
The extension was _still_ marked as LISTED/VISIBLE, of course they block it.

I marked it as unlisted now and I uploaded a new version manually. It's still awaiting review but according to your previous review this is allowed:

![image](https://user-images.githubusercontent.com/37769974/74526654-d56b3000-4f49-11ea-83d4-29a2278792ad.png)

"you have to sign it and self-host"

fregante commented:
Indeed they just approved it, however for some reason `web-ext sign` takes ages, probably never ending. This should be instant 🤦‍♂️ 

```
$ npx web-ext sign --source-dir distribution
Applying config file: ./package.json
Building web extension from ./distribution
Validating add-on [........................................................................................]
Validation results: https://addons.mozilla.org/en-US/developers/upload/15ee7c54221946ada5ba8303cd9c33a9
Signing add-on [...........................................................................................]
Downloading signed files: 100% 
Downloaded:
    ./web-ext-artifacts/awesome_extension-20.3.25.3-an+fx.xpi
Extension ID: awesome-extension@notlmn.github.io
SUCCESS

```

Edit: it took 4 minutes but it worked. This should be fine 👌

fregante commented on package.json:
```diff
@@ -25,7 +25,6 @@
 		"terser-webpack-plugin": "^1.3.0",
-		"utc-version": "^2.0.1",
 		"web-ext": "^3.1.1",
 		"web-ext-submit": "^3.1.1",
```
```suggestion
		"web-ext": "^4.1.0",
		"web-ext-submit": "^4.1.0",
```

fregante committed:
Cleanup/update

notlmn commented:
> The extension was still marked as LISTED/VISIBLE, of course they block it.

I didn't explicitly mark it as "unlisted" because the description sounded that it would disable the extension completely. My bad, should've tried that option too. 😅

---------

> however for some reason `web-ext sign` takes ages, probably never ending. This should be instant

Just created a new automated version release locally, it was instant, and Mozilla seems to auto validate if the extension is unlisted, was approved instantly too through "automatic validation".

```
$ npm run release:amo

> @ release:amo /mnt/d/dev/git/notlmn/brower-extension-template
> web-ext-submit --source-dir distribution

Applying config file: ./package.json
Building web extension from /mnt/d/dev/git/notlmn/brower-extension-template/distribution
Validating add-on [............................................................]
Validation results: https://addons.mozilla.org/en-US/developers/upload/63a940bbcefb4975bfbffb2cfe9315ec
Downloading signed files: 100%
Downloaded:
    ./web-ext-artifacts/awesome_extension-20.3.25.4-an+fx.xpi
Extension ID: awesome-extension@notlmn.github.io
SUCCESS
```

--------------

Anyway, it is working as intended, and should be good to go.

fregante committed:
Update readme

fregante committed:
Change cronjob from daily to weekly

fregante commented:
Docs updated.

Before merging we just need to clear those TODOs

fregante committed:
Improve auto-deploy comments and drop TODO

fregante committed:
Rename deploy-on-version-tag.yml to deploy-tag.yml

fregante committed:
Drop TODOs from deploy-tag.yml

fregante committed:
Only allow hotfixes via manual tagging

notlmn commented on readme.md:
```diff
@@ -156,39 +156,20 @@ npm run release
 ### License
 
-This browser extension template is released under [MIT](#license) and mentioned below. There is no `license` file included in here, but when you clone this template, you should include your own license file for the specific license you choose to use.
+This browser extension template is released under [CC0](#license) and mentioned below. There is no `license` file included in here, but when you clone this template, you should include your own license file for the specific license you choose to use.
```
# ♥

notlmn commented on .github/workflows/deploy-automatic.yml:
```diff
@@ -0,0 +1,21 @@
+
+on:
+  schedule:
+    - cron: '13 13 * * 3'
```
```suggestion
    - cron: '31 13 * * 3' # Every Wednesday at 13:31 https://crontab.guru/#31_13_*_*_3
```

notlmn committed:
Change cron to `13:31`

fregante commented on .github/workflows/deploy-tag.yml:
```diff
@@ -0,0 +1,56 @@
+    - run: npm test
+    - name: Version and tag
+      run: |
+        export VER=$(echo "$GITHUB_REF" | sed -e 's/refs\/heads\///')
```
This isn't working as expected: https://github.com/npmhub/npmhub/runs/540747564#step:3:9

Generated `manifest.json`

<img width="396" alt="manifest" src="https://user-images.githubusercontent.com/1402241/77806367-c87f4780-7084-11ea-9391-8972f357ed9b.png">


On my computer it does:

```sh
bash-3.2$ GITHUB_REF=refs/heads/feature-branch-1
bash-3.2$ VER=$(echo "$GITHUB_REF" | sed -e 's/refs\/heads\///')
bash-3.2$ echo $VER
feature-branch-1
```

notlmn replied:
Looks like `GITHUB_REF` is in the form `refs/tags/TAG` and not `refs/heads/TAG`.

fregante committed:
Fix deployment workflow

fregante commented:
Woot! Firefox deployment confirmed to work!

Tag created manually on GitHub: https://github.com/npmhub/npmhub/releases/tag/2020.3.28
Successful deployment: https://github.com/npmhub/npmhub/runs/542600369?check_suite_focus=true
Version on AMO (currently in review):
<img width="216" alt="amo" src="https://user-images.githubusercontent.com/1402241/77838384-b16b5300-7173-11ea-9306-be8811368634.png">

notlmn commented:
🎉 Thank you @fregante.
